SNAP/Kokkos work batching and instruction-level parallelism optimizations

weinbe2 opened the pull request:
**Summary**

This PR introduces work batching within the core kernels in the Kokkos implementation of SNAP: `ComputeUi`, `ComputeYi`, and `ComputeFusedDeidrj`. The batching of work is designed to increase instruction-level parallelism, in particular for fp64 computations, relative to other work (integer math, lookup table loads, atomic addition) as appropriate for each kernel. Each kernel has its own new batching parameter available at compile time, and by design a batch value of 1 makes the kernels equivalent to their initial implementation.

* ComputeUi: batching quantity `ui_batch`, parallelism is promoted from (atom, neighbor) to (atom, `ui_batch` neighbors), reducing atomic additions by a factor of `ui_batch`
* ComputeYi: batching quantity `yi_batch`, parallelism is promoted from (atom, idxz) to (`yi_batch` atoms, idxz), reducing lookup table loads and some integer arithmetic
* ComputeFusedDeidrj: fuses all three directions (controlled by a boolean `use_deidrj_all`), reducing redundant work computing `u` values in the product rule and loading `y` values when computing the force

Of note, these values are set on a per-architecture basis near the top of `pair_snap_kokkos.h` and are propagated into `SNAKokkos` routines via template parameters. Tuned values were provided for the NVIDIA architecture and "default" values that should cause performance to reduce to that of the original code were set for other architectures. I believe that there should be tunable values for other architectures that will lead to performance boosts (exposing proportionally more fp64 work should benefit all), I just don't have the means to test it myself.

_Note:_ I had to make minimal changes to the SNAP grid files since they called routines in `SNAKokkos`, setting the default values. I would appreciate it if someone could do a quick _correctness_ and sanity check.

**Related Issue(s)**

Iterative improvement on #2599 , #2905 , #3465 , #4391 , other Kokkos/SNAP PRs...

**Author(s)**

Evan Weinberg (NVIDIA). eweinberg [at] nvidia.com preferred, evansweinberg [at] gmail.com for "long-lived".

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No change, though I did change the default value for the cross-over from the "small" atom count parallelism to "large" atom count from 8192 to 6144 to avoid a step-function effect due to extra padding for 8000 atoms, which may cause a change in performance for atom count values between ~6000 and ~8000 at the default batch sizes.

**Implementation Notes**

The core refactoring was replacing various scalar variables with `Kokkos::Array` of those variables, and then replacing the calculations using the scalar variables with `#pragma unroll`'d parallel for expressions over the array elements. This is exposed by a `register_loop` abstraction. For example, code that originally looks like:

```
complex a, b, ulist_prev, ulist_accum;
real_type rootpq;
[...]
ulist_accum.re += rootpq * (a.re * ulist_prev.re + a.im * ulist_prev.im);
ulist_accum.im += rootpq * (a.re * ulist_prev.im - a.im * ulist_prev.re);
```

becomes

```
Kokkos::Array<complex, ui_batch> a, b, ulist_prev, ulist_accum;
real_type rootpq;
[...]
register_loop<ui_batch>([&] (int n) -> void {
  ulist_accum[n].re += rootpq * (a[n].re * ulist_prev[n].re + a[n].im * ulist_prev[n].im);
  ulist_accum[n].im += rootpq * (a[n].re * ulist_prev[n].im - a[n].im * ulist_prev[n].re);
});
```

Where the details of the `for` loop, unrolling, etc, is hidden within `register_loop`. A `ui_batch` value of "1" makes the code functionally equivalent to the original code.

This code was verified for correctness and was confirmed to give performance improvements on V100-16GB, A100-80GB, and H200-HBM3e. This was verified for linear SNAP at 2J8 and 2J14, along with pSNAP and for qSNAP (using the Gordon Bell carbon dataset). The values used were `ui_batch` (4), `yi_batch` (4), and `use_deidrj_all` (true).

The one exception to performance boosts was, of course, the ECP problem at 2J14 uniquely on V100-16GB where I recorded a 4% regression. I imagine this could be tuned around by modifying one or more of the compile-time parameters, possibly with a conditional on Volta using, for ex, `#ifdef KOKKOS_ARCH_VOLTA`.

As noted above, I believe that there should be non-default values for one or more of `ui_batch`, `yi_batch`,and `use_deidrj_all` that lead to performance boosts for other GPUs. These could either be modified in a blanket form for HIP and SYCL, or at finer-grained detail using more specific preprocessor defines like `KOKKOS_ARCH_AMD_GFX940` vs `KOKKOS_ARCH_AMD_GFX942`, etc.

I've included a table of performance and speed-up data on H200-HBM3e below, obtained using CUDA 12.8.0. I've included similar tables for V100 and A100, along with a painfully exhaustive set of data across atom counts for H200, in the further information section.

All performance numbers are in units of m-atom timesteps / second.

| Model             | Atoms  | Develop | Optimized | Performance Boost |
|-------------------|--------|---------|-----------|-------------------|
| Linear SNAP 2J8   | 2000 | 2.297 | 2.886 | 1.26x |
| Linear SNAP 2J8   | 512000 | 4.296 | 6.456 | 1.50x |
| Linear SNAP 2J14  | 2000   | 0.352 | 0.446 | 1.27x |
| chemsnap          | 512000 | 2.549 | 3.411 | 1.34x |
| qSNAP             | 373248 | 3.123 | 4.251 | 1.36x |

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system

**Further Information, Files, and Links**

Extra perf data, units m-atom timesteps / second:

V100:
| Model             | Atoms  | Develop | Optimized | Performance Boost |
|-------------------|--------|---------|-----------|-------------------|
| Linear SNAP 2J8   | 2000   | 1.143   | 1.245     | 1.09x             |
| Linear SNAP 2J8   | 512000 | 1.602 |1.933 | 1.21x |
| Linear SNAP 2J14  | 2000   | 0.139 | 0.133   | 0.96x             |
| chemsnap          | 512000 | 0.998109 | 1.187	| 1.19x |
| qSNAP             | 373248 | 1.224 | 1.369 | 1.12x |


A100:
| Model             | Atoms  | Develop | Optimized | Performance Boost |
|-------------------|--------|---------|-----------|-------------------|
| Linear SNAP 2J8   | 2000   | 1.43 | 1.706 | 1.19x |
| Linear SNAP 2J8   | 512000 | 2.145 | 2.855 | 1.33x |
| Linear SNAP 2J14  | 2000 | 0.183 | 0.220 | 1.20x |
| chemsnap          | 512000 | 1.408 | 1.792 | 1.27x |
| qSNAP             | 373248 | 1.667 | 2.117 | 1.27x |

Tables of data for H200:

Table 1: Linear SNAP 2J8, units m-atom timesteps/second
| Atoms  | Develop | Optimized | Performance Boost |
|--------|---------|-----------|-------------------|
| 2000   | 2.297   | 2.886     | 1.26x             |
| 4000   | 2.972   | 3.883     | 1.31x             |
| 8000   | 3.386   | 4.659     | 1.38x             |
| 16000  | 3.757   | 5.455     | 1.45x             |
| 32000  | 4.027   | 5.939     | 1.47x             |
| 64000  | 4.151   | 6.217     | 1.50x             |
| 128000 | 4.235   | 6.351     | 1.50x             |
| 256000 | 4.272   | 6.413     | 1.50x             |
| 512000 | 4.296   | 6.456     | 1.50x             |
| 1024000| 4.31    | 6.464     | 1.50x             |
| 2048000| 4.314   | 6.467     | 1.50x             |

Table 2: Linear SNAP 2J14, units k-atom timesteps/second
| Atoms  | Develop  | Optimized | Performance Boost |
|--------|----------|-----------|-------------------|
| 2000   | 351.649  | 445.664   | 1.27x             |
| 4000   | 361.163  | 470.273   | 1.30x             |
| 8000   | 368.839  | 475.427   | 1.29x             |
| 16000  | 368.843  | 496.461   | 1.35x             |
| 32000  | 375.588  | 505.056   | 1.34x             |
| 64000  | 378.318  | 503.113   | 1.33x             |
| 128000 | 378.438  | 503.829   | 1.33x             |

Table 3: chemsnap, units m-atom timesteps/second
| Atoms  | Develop | Optimized | Performance Boost |
|--------|---------|-----------|-------------------|
| 1000   | 1.117   | 1.277     | 1.14x             |
| 2000   | 1.551   | 1.818     | 1.17x             |
| 4000   | 1.916   | 2.401     | 1.25x             |
| 8000   | 2.176   | 2.94      | 1.35x             |
| 16000  | 2.439   | 3.21      | 1.32x             |
| 32000  | 2.564   | 3.422     | 1.33x             |
| 64000  | 2.522   | 3.371     | 1.34x             |
| 128000 | 2.542   | 3.397     | 1.34x             |
| 256000 | 2.536   | 3.367     | 1.33x             |
| 512000 | 2.549   | 3.411     | 1.34x             |
| 1024000| 2.566   | 3.373     | 1.31x             |

Table 4: qSNAP, units m-atom timesteps/second
| Atoms  | Develop | Optimized | Performance Boost |
|--------|---------|-----------|-------------------|
| 46656  | 3.08    | 4.177     | 1.36x             |
| 93312  | 3.11    | 4.191     | 1.35x             |
| 186624 | 3.133   | 4.238     | 1.35x             |
| 373248 | 3.123   | 4.251     | 1.36x             |
| 746496 | 3.157   | 4.307     | 1.36x             |
| 1492992| 3.137   | 4.254     | 1.36x             |

weinbe2 committed:
Added support for extra padding to data structures to simplify extra vectorization

weinbe2 committed:
Added register tiling to ComputeYi/Zi/YiWithZlist

weinbe2 committed:
Added register tiling to ComputeUi

weinbe2 committed:
Added directional fusion to ComputeFusedDeidrj

weinbe2 committed:
Added register tiling to ComputeBi

weinbe2 committed:
Reduced template parameters, improved code reuse between ComputeFusedDeidrj paths

weinbe2 committed:
Added compile-time support for running the legacy codepaths on the GPU; misc cleanup and fixes

weinbe2 committed:
Fixed up whitespace

weinbe2 committed:
nvc++ compiler workaround

weinbe2 committed:
Merge branch 'develop' into feature/batch-snap-ilp

weinbe2 committed:
Merge branch 'develop' into feature/batch-snap-ilp

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into feature/batch-snap-ilp

stanmoore1 committed:
Small tweak from @cjknight for Aurora performance

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into feature/batch-snap-ilp

stanmoore1 commented:
I'm seeing a bunch of these warnings with HIP:
```
../sna_kokkos.h:89:21: warning: argument to '#pragma unroll' should not be in parentheses in CUDA C/C++ [-Wcuda-compat]
   89 |     #pragma unroll (num_elems)
```

stanmoore1 committed:
Fix warning by removing parentheses

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into feature/batch-snap-ilp

stanmoore1 committed:
Tuning for AMD GPUs

stanmoore1 commented:
I tested this on AMD MI300A and I do see a significant speedup after tuning, but not as large as NIVIDIA H100.

stanmoore1 commented:
This PR gives nearly 1.5x speedup on MI250X using the same params as for MI300A!

stanmoore1 commented on src/KOKKOS/compute_sna_grid_kokkos.h:
```diff
@@ -249,6 +255,7 @@ class ComputeSNAGridKokkos : public ComputeSNAGrid {
   double lo0, lo1, lo2;
 
   // Make SNAKokkos a friend
+  // the 1 is because ComputeSNAGridKokkos does not support padded vectorization yet
```
@weinbe2 I don't understand this comment, where is the "1"?

weinbe2 replied:
That's a legacy comment from some intermediate implementations that I rolled back, I can delete it.

weinbe2 replied:
https://github.com/weinbe2/lammps/commit/625811cc11f67abae4467afe30a26329ab2dd825

stanmoore1 commented on src/KOKKOS/sna_kokkos_impl.h:
```diff
@@ -540,35 +597,45 @@ void SNAKokkos<DeviceType, real_type, vector_length>::evaluate_ui_jbend(const Wi
-  ulist_wrapper.set(0, complex::one());
+  // level 0 is just 1; it's now pre-initialized
+  //{
+  //  Kokkos::Array<complex, ui_batch> ulist_accum;
```
Is this commented out code still needed, e.g. for reference?

weinbe2 replied:
It's there to note that I moved a detail of data initialization elsewhere, I'll make this clearer.

weinbe2 replied:
https://github.com/lammps/lammps/pull/4601/commits/65f8f8695fd6fe9229b31e5b9d9601361d5d4872

stanmoore1 reviewed (approved):
I approve, but had 2 questions about some comments in the code.

weinbe2 committed:
Clean up an intermediate comment in `compute_sna_grid_kokkos.h`

weinbe2 committed:
Added documentation on initializing `ulist_wrapper` in `evaluate_ui_jbend`

weinbe2 committed:
Added self to contributors list (a few years late)

weinbe2 committed:
Added various documentation about the `*_batch` values to `pair_snap_kokkos.h`

weinbe2 commented:
@stanmoore1 I think I addressed your comments, and I also added some documentation that covers questions that came up in some of our offline conversations. Let me know if you'd like any further modifications!

Also, thank you for tuning `ui_batch`, `yi_batch`, and `use_deidrj_all` for MI250X and MI300A! I'm happy my optimizations can give a real performance improvement across multiple architectures. (Similarly, thank you @cjknight for looking into Aurora performance!)

weinbe2 committed:
Merge branch 'develop' into feature/batch-snap-ilp

stanmoore1 committed:
whitespace

stanmoore1 committed:
Merge branch 'feature/batch-snap-ilp' of github.com:weinbe2/lammps into feature/batch-snap-ilp
